test/e2e/network/service.go

```diff
@@ -4226,7 +4227,81 @@ var _ = common.SIGDescribe("Services", func() {
 		service, err := t.CreateService(service)
 		framework.ExpectNoError(err)
 
-		checkServiceReachabilityFromExecPod(ctx, f.ClientSet, ns, service.Name, service.Spec.ClusterIP, port)
+		_ = checkServiceReachabilityFromExecPod(ctx, f.ClientSet, ns, service.Name, service.Spec.ClusterIP, port)
+	})
+
+	ginkgo.It("should connect to the named ports during port number changing", func(ctx context.Context) {
+		serviceName := "mutable-named-port"
+		ns := f.Namespace.Name
+
+		t := NewServerTest(cs, ns, serviceName)
+		defer func() {
+			defer ginkgo.GinkgoRecover()
+			errs := t.Cleanup()
+			if len(errs) != 0 {
+				framework.Failf("errors in cleanup: %v", errs)
+			}
+		}()
+
+		port := int32(8080)
+		portName := "mutable-port"
+		ports := []v1.ContainerPort{{Name: portName, ContainerPort: port, Protocol: v1.ProtocolTCP}}
+		args := []string{"netexec", fmt.Sprintf("--http-port=%d", port)}
+		pods := []*v1.Pod{}
+		for _, podName := range []string{"testpod0", "testpod1"} {
+			pods = append(pods, createPodOrFail(ctx, f, ns, podName, t.Labels, ports, args...))
+		}
+
+		service := t.BuildServiceSpec()
+		servicePort := int32(8001)
+		service.Spec.Ports = []v1.ServicePort{
+			{
+				Name:       portName,
+				Port:       servicePort,
+				TargetPort: intstr.FromString(portName),
+			},
+		}
+
+		ginkgo.By(fmt.Sprintf("creating Service %v with selectors %v", service.Name, service.Spec.Selector))
+		service, err := t.CreateService(service)
+		framework.ExpectNoError(err)
+
+		getPodNamesBehindService := func(expectPodCnt int) (podNames []string) {
+			for retryCnt := 0; retryCnt < max(20, expectPodCnt*2); retryCnt++ {
+				output := checkServiceReachabilityFromExecPod(ctx, f.ClientSet, ns, service.Name, service.Spec.ClusterIP, servicePort)
+				if !slices.Contains(podNames, output) {
+					podNames = append(podNames, output)
+				}
+			}
+			slices.Sort(podNames)
+			return podNames
+		}
+
+		hostnames := getPodNamesBehindService(len(pods))
+		if !slices.Equal(hostnames, []string{"testpod0", "testpod1"}) {
+			framework.Failf("Failed to check pod connectivity for service %s/%s with session affinity, got pod names %v", ns, serviceName, hostnames)
+		}
+
+		ginkgo.By(fmt.Sprintf("delete pod %s and recreate with the same port number under port name %s", pods[0].Name, portName))
```

Comment:
What we want is to first delete testpod0 and add a new testpod2 using the _new_ port, then delete testpod1 and add testpod3 also using the new port. We don't need to test recreating pods with the old port.